### PR TITLE
KE2: Extract lambda expressions

### PR DESCRIPTION
**Pull request by tamasvajk**

### CODE REVIEW

File: java/kotlin-extractor2/src/main/kotlin/entities/FunctionalInterface.kt [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+/** Extracts the class around a local function, a lambda, or a function reference. */
+context(KaSession)
+@OptIn(KaExperimentalApi::class)
+private fun KotlinFileExtractor.extractGeneratedClass(
+    ids: GeneratedClassLabels,
+    superTypes: List<KaType>,
+    locId: Label<DbLocation>,
+    elementToReportOn: PsiElement,
+    compilerGeneratedKindOverride: CompilerGeneratedKinds? = null,
+    superConstructorSelector: (KaFunctionSymbol) -> Boolean = { it.valueParameters.isEmpty() },
```
Comment (igfoo) [inline java/kotlin-extractor2/src/main/kotlin/entities/FunctionalInterface.kt:452]: It looks like we never override this in KE1; should we just inline the default?

File: java/kotlin-extractor2/src/main/kotlin/entities/FunctionalInterface.kt [hunk 2]
```diff
@@ -0,0 +1,569 @@
...
+    extractSuperConstructorArgs: (Label<DbSuperconstructorinvocationstmt>) -> Unit = {},
```
Comment (igfoo) [inline java/kotlin-extractor2/src/main/kotlin/entities/FunctionalInterface.kt:453]: Can we just inline the default here too?

File: java/kotlin-extractor2/src/main/kotlin/entities/FunctionalInterface.kt [hunk 3]
```diff
@@ -0,0 +1,569 @@
...
+
+/** Extracts the class around a local function, a lambda, or a function reference. */
+context(KaSession)
+@OptIn(KaExperimentalApi::class)
+private fun KotlinFileExtractor.extractGeneratedClass(
+    ids: GeneratedClassLabels,
+    superTypes: List<KaType>,
+    locId: Label<DbLocation>,
+    elementToReportOn: PsiElement,
+    compilerGeneratedKindOverride: CompilerGeneratedKinds? = null,
```
Comment (igfoo) [inline java/kotlin-extractor2/src/main/kotlin/entities/FunctionalInterface.kt:451]: If this is a generated class, then is the `null` default ever correct? Should we just make it a required parameter?

File: java/kotlin-extractor2/src/main/kotlin/entities/Expression.kt [hunk 4]
```diff
@@ -1353,6 +1268,16 @@ private fun KotlinFileExtractor.extractExpression(
                                 )
                             }
             */
+
+            is KtLambdaExpression -> {
+                // Propagate extraction to the wrapped function literal:
+                return extractExpression(e.functionLiteral, callable, parent)
```
[the remark this answers is not in the document]
Comment (tamasvajk) [reply java/kotlin-extractor2/src/main/kotlin/entities/Expression.kt:1274]: I think so.

File: java/kotlin-extractor2/src/main/kotlin/entities/FunctionalInterface.kt (hunk 2 above)
Comment (igfoo) [reply java/kotlin-extractor2/src/main/kotlin/entities/FunctionalInterface.kt:452]: I think this can now be removed

Comment (igfoo) [reply java/kotlin-extractor2/src/main/kotlin/entities/FunctionalInterface.kt]: Ah, sorry, I missed that. Anyway, like you say, we can always add it back if we need it.

**Review by igfoo — APPROVED** [review]

**Pull request merged**